### DISCUSSION TITLE
EGOS for Pine64 Ox64

### DISCUSSION THREAD

**Discussion by ThothK** [yhzhang0128/egos-2000#7] [General]
Hello,

	I am very interesting in a port to the Pine64 Ox64, information for which is located here:
	https://pine64.com/product/128mb-ox64-sbc-available-on-december-2-2022/
	Some discussion and component details are located here:
	https://wiki.pine64.org/wiki/Ox64

	The important components as I understand them are:
	Bouffalo Lab BL808 RISC-V SoC
		64-bit 480MHz RV64 C906
		32-bit 320MHz RV32 E907
		32-bit 150MHz RV32 E902
	Memory
		728KB internal SRAM
		64MB internal PSRAM
		128Mb (16MB) XSPI Flash
		MicroSD slot
	I/O Interface
		UART
		I2C
		SPI
		I2S
	Communication
		802.11 b/g/n 2.4GHz Wifi
		BLE 5.0
		Zigbee
		USB
		10/100Mbps Ethernet thru optional add-on board

	I have read the comments of the 'issue' regarding porting to the ESP32 C3.

	I would like a rough plan describing what it would take to port EGOS to the 0x64.
	Depending on the amount of effort required, I could do this. 
	There is even a possibility of allowing my employer to give me time to do some of the work.

	While I am an experienced C/C++ developer, I have little experience with bare metal programming (as I understand it).
	In addition to learning what I do not know, I would like assistance determining what I would need to know.

	To start:
	From the ESP32C3 discussion I saw

	"The key is to adjust the memory regions for this processor."
	"Put the earth layer to the 384 KB ROM
	Put the 4MB disk image to the embedded flash
	Adjust other memory regions into the 400 KB SRAM."
	The above would need to be adjusted for the Ox64, but what should go where?

	"In order to do this, you will need to read and modify All the lds files in this repo"
	What are the lds files for?

	"The memory map defined here: https://github.com/yhzhang0128/egos-2000/blob/main/library/egos.h#L53"

	Is there a particular need to use the SiFive RISCV compiler (SiFive riscv-gcc)?
	Recent versions of both GCC and LLVM / Clang have basic RISC-V support.

**Comment by yhzhang0128** [comment]
Thank you for your interest in porting it to 0x64. 

It is hard for me to give you a full plan right now, but a good starting point is to do two things: (1) run a simple `printf("Hello, World!\n");` program on 0x64; (2) test interrupt handling on 0x64. They corresponds to `cpu_intr.c` and `dev_tty.c` under the `earth` directory.

After these two steps, you will be at a good stage for modifying the memory layout and doing the related debugging. And we can discuss here how to proceed with it. And we can use this discussion thread for Q&A on low-level programming as well.

`lds` is a script file for the linker -- telling the compiler which address to put code section, data section, heap section, etc. You may take a look at one of them and its format seems easy to understand.

I use the SiFive compiler mostly because it integrates `Newlib`, a C library implementation. For example, there is no `printf` in my code because my code is linked with the `printf` implemented in `Newlib`. If other compiler support something similar, we don't have to use the one from SiFive.
